Async using callbacks;

grsin opened the pull request:
Tested using:
```
const newwalk = require('./src/walk-manifest.js');
const oldwalk = require('../../hls-fetcher/src/walk-manifest');
const assert = require('assert');

const urls = ['http://qthttp.apple.com.edgesuite.net/1010qwoeiuryfg/sl.m3u8',
  'http://devimages.apple.com/iphone/samples/bipbop/bipbopall.m3u8',
  'http://devimages.apple.com/iphone/samples/bipbop/gear1/prog_index.m3u8',
  'http://playertest.longtailvideo.com/adaptive/oceans_aes/oceans_aes.m3u8',
  'http://playertest.longtailvideo.com/adaptive/captions/playlist.m3u8',
  'http://playertest.longtailvideo.com/adaptive/wowzaid3/playlist.m3u8',
  'http://content.jwplatform.com/manifests/vM7nH0Kl.m3u8',
  'http://cdn-fms.rbs.com.br/hls-vod/sample1_1500kbps.f4v.m3u8',
  'http://sample.vodobox.net/skate_phantom_flex_4k/skate_phantom_flex_4k.m3u8',
  'https://bitdash-a.akamaihd.net/content/MI201109210084_1/m3u8s/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.m3u8',
  'https://bitdash-a.akamaihd.net/content/sintel/hls/playlist.m3u8',
  'https://mnmedias.api.telequebec.tv/m3u8/29880.m3u8'];

urls.forEach(function(url) {
  console.log('Checking ' + url);
  newwalk(false, '.', url, function(err, newResources) {
    const oldResources = oldwalk(false, '.', url);
    assert.deepStrictEqual(newResources, oldResources, 'NO Decrypt | ASYNC VS SYNC should be the same')
  });

  newwalk(true, '.', url, function(err, newResources) {
    const oldResources = oldwalk(true, '.', url);
    assert.deepStrictEqual(newResources, oldResources, 'Decrypt |ASYNC VS SYNC should be the same')
  });
});
```

Also I couldn't tell what exact style you guys were following; but I guess as mentioned in a different PR - linting will get added so maybe we lint it there? Or I can change it back to just match as much as possible as before.

mwhipple commented on src/cli.js:
```diff
@@ -27,11 +27,13 @@ var options = {
-}).catch(function(error) {
-  console.error('ERROR', error);
-  process.exit(1);
+start(options, function(data) {
```
so start is passed a callback which will be passed a promise?

grsin replied:
AH wrong file! I'll fix that.

gkatsev replied:
since WriteData uses bluebird for promises, I think we use it's [`asCallback`](http://bluebirdjs.com/docs/api/ascallback.html) method to have it work as either one. That way, WriteData internally can use promises but its external interface can be node style callbacks.

mwhipple commented on test/unit/walk-manifest.spec.js:
```diff
@@ -0,0 +1,101 @@
+
+            walker(false, '.', TEST_URL + '/test.m3u8', function(err, resources) {
+                // 4 m3u8 and 8 * 3 segments
+                assert.equal(resources.length, 28);
```
probably uniquify this list if we're leaning on pigeonhole

BrandonOCasey commented on package.json:
```diff
@@ -24,12 +24,18 @@
     "promise-streams": "^1.0.1",
+    "request": "^2.87.0",
     "requestretry": "^1.8.0",
     "sync-request": "^3.0.1"
```
we can probably remove `sync-request`

mwhipple commented on test/unit/walk-manifest.spec.js:
```diff
@@ -0,0 +1,282 @@
+            });
+        });
+
+        it('should return not break if sub playlist is not a valid m3u8', function(done) {
```
shouldn't errors also result in errors?

mwhipple commented on test/unit/walk-manifest.spec.js:
```diff
@@ -0,0 +1,282 @@
+            });
+        });
+
+        it('should not  for a cycle m3u8', function(done) {
```
the better test for a cycle would be two playlists that reference each other.

mwhipple commented on test/unit/walk-manifest.spec.js:
```diff
@@ -0,0 +1,282 @@
+            });
+        });
+
+        it('should return not throw an error if one sub playlist gets an error', function(done) {
```
that seems like it should be configurable behavior. Is this also asserting that the errors can be made visible?

mwhipple commented on package.json:
```diff
@@ -24,12 +24,18 @@
     "mkdirp": "^0.5.1",
     "pessimist": "^0.3.5",
     "promise-streams": "^1.0.1",
+    "request": "^2.87.0",
```
should this also have a browser compatible equivalent? (not sure how that stuff fits together)

gkatsev replied:
are we trying to make this module work in the browser as well as in node?

mwhipple replied:
doesn't it currently? We don't care but I was just concerned about regression.

BrandonOCasey replied:
I don't think the intent was ever for this to run in the browser. only in node.

mwhipple replied:
cool, n/m then 🙈

forbesjo replied:
Right, the project's not set up to work in the browser

gkatsev commented on src/index.js:
```diff
@@ -1,12 +1,12 @@
-  return WriteData(options.decrypt, options.concurrency, resources);
+  WalkManifest(options.decrypt, options.output, options.input, function(err, resources) {
+    console.log("Downloading additional data...");
+    callback(WriteData(options.decrypt, options.concurrency, resources));
```
should callback be a node-style callback? Also, if `err` is an error object, we probably don't want to call `WriteData`.

gkatsev commented on src/walk-manifest.js:
```diff
@@ -49,116 +50,150 @@ var parseManifest = function(content) {
+          return cb(null);
+        }
+        walkPlaylist(decrypt, basedir, p.uri, manifest, playlists.indexOf(p), cb);
+      }, function(err, results) {
```
what happens if there are errors?

gkatsev commented on src/walk-manifest.js:
```diff
@@ -49,116 +50,150 @@ var parseManifest = function(content) {
+  request({url: keyUri, encoding: null, timeout: 1500}, function(error, response, body) {
+    if (error) {
+      console.error('Failed to get key', error, keyUri);
+      return callback({});
```
it'll be good to make all callback node-style callbacks.

grsin commented:
So a few new things.

* continueOnError - There is a new option if true will keep going continue recursion even if an error is found for a sub playlist (or a sub^n playlist). All those errors will get added to the resources array. Now that is kinda weird; I kinda had to do this since I can't really call back with an error otherwise it will break. Resources will be a mix of valid resources and error objects. So the caller will not get err but will have to parse the resources array. So either we let the caller do the filtering or we add a wrapper function that will filter out all the errors and the valid resources and callback with an array of errors and and array of valid resources. That feels kinda weird to me since normally if you get an err in a callback you normally don't look at the rest of the results but eh.

Something like 
```
var walkPlaylist = function(decrypt, basedir, uri, parent, manifestIndex, continueOnError, callback) {
  wk(decrypt, basedir, uri, parent, manifestIndex, continueOnError, function(err, resources) {
    if (!continueOnError) {
      callback(err);
    } else {
      const nestedErrors = resources.filter(e => e instanceof Error);
      const validResources = resources.filter(r => 'uri' in r);
      callback(nestedErrors, validResources);
    }
  });
};
```

* Added cycle detection so 2 playlists can't keep referencing each other and we get stuck in an infinite loop. I don't know if this is really possible but eh for some manifests we get we don't own 100% so if someone does something weird don't break. I added clearVisited so we could clear the cycle checker list after each run.

* Promises. I got a promises version working [here](https://github.com/grsin/hls-fetcher/pull/1). Should be pretty easy to convert over if we wanted; most of the above will still be the case for that version too. I just did the callback way since I'm more familiar with it and wanted to finish the test cases; I don't really mind either way if folks want me to convert it over.

mwhipple commented on src/walk-manifest.js:
```diff
@@ -49,116 +58,192 @@ var parseManifest = function(content) {
+        reflectedResults.forEach(function(r) {
+          if (r.error && !continueOnError) {
+            console.error('Exiting early continueOnError false...|' + manifest.uri);
+            callbackTopError = r.error;
```
So is the setup here that all of the sub playlists will always be processed regardless of the continueOnError value and then the code here will effectively promote errors that exist if continueOnError is false?

grsin replied:
Ya. on continueOnError is false the sublist will keep running but will just error early and not continue to recurse if they fail.

mwhipple replied:
pulling out continueOnError into the strategy style could avoid the need for this. The idea of bypassing behavior to later emulate it seems suspect.

mwhipple commented on src/walk-manifest.js:
```diff
@@ -1,27 +1,36 @@
+
+var clearVisited = function() {
+  visitedUrls = []
+  return;
```
An alternative approach that would bake cycle protection into the logic (and also allow switching to tail calls which can be optimized in ES6) would be to accumulate the segments in an object of {$url: [$segments]} in the inner functions where the values are then flattened to be returned for the existing signature.

grsin commented on src/walk-manifest.js:
```diff
@@ -49,116 +58,192 @@ var parseManifest = function(content) {
+      return callback(keyError, {});
+    }
+    if (response.statusCode !== 200) {
+      const keyError = new Error(response.statusCode + '|' + keyUri);
```
Should we add the response body too?

mwhipple replied:
I'd think the entire response could be passed as part of the callback error and maybe logged at a lower log level (if that makes sense with logging system). Special handling may also make sense fro 404s. I'd also be interested in what the behavior around redirects is (presumably the client will follow them so handling them shouldn't be expected).

grsin replied:
We could pass the response but that seems like a lot to pass to the callback. That could be a lot of object/info passed. It seems like it's enough if we see an error we just make a simple error object and then you can debug later if you want maybe - just try to walk as best effort and if some url or manifest is bad that's ok since it will get expired out the cache soon enough. I think all the status codes will be just visual checks - if you get a non 200 that means the client won't be able to play it back so we don't really care. We only care about the 200s since we really only need to purge those. Redirects are just handled by request they will follow (up to a point).

mwhipple replied:
The data passed to the callback would just be a reference to an object that already exists. Passing the full object allows the client code to decide how it wants to handle the errors which allows this library to be slightly more generalized rather than being biased towards specific consumers. Non-200s are not necessarily hard failures: from what I remember one of the major causes of issues in similar processing I've done in the past was access issues revolving around things like not including a proper token param for CDN access which is something that some calling code may be capable of and want to handle. It seems like providing a tailored projection of failure data could be helpful but dropping the rest of the data seems to needlessly limit the ways the library could be used.

grsin commented on src/walk-manifest.js:
```diff
@@ -49,116 +58,192 @@ var parseManifest = function(content) {
+    parent = false;
+    manifestIndex = 0;
+    continueOnError = false;
+  }
```
So I'm not 100% sure how to handle default parameters with callbacks (even ES6 default doesn't work nicely here). I could do a check to see if one of the parameters is a function (callback) but I thought based on the way we use walk this should be ok.

gkatsev replied:
I think it's time to change this method to just take an options object instead.
```js
var WalkPlaylist = function(options, callback) {
  var {
    decrypt, 
    basedir,
    uri,
    parent = false,
    manifestIndex = 0,
    continueOnError = false
  } = options;
```

gkatsev commented on src/walk-manifest.js:
```diff
@@ -49,116 +58,192 @@ var parseManifest = function(content) {
+    if (error) {
+      const keyError = new Error(error.message + '|' + keyUri);
+      console.error(keyError, error);
+      return callback(keyError, {});
```
no need to pass a second argument when we have an error, calling code shouldn't assume that data was provided if an error was passed.

mwhipple commented on src/walk-manifest.js:
```diff
@@ -49,116 +50,197 @@ var parseManifest = function(content) {
+  request({url: manifest.uri, timeout: 1500}, function(error, response, body) {
+    if (error) {
+      const manifestError = new Error(error.message + '|' + manifest.uri);
+      return onError(manifestError, resources, callback);
```
I'd think for both of these passing the richer object through to the error handler in form or another may be better, especially so the handler can branch without having to do string parsing.

mwhipple commented on src/walk-manifest.js:
```diff
@@ -49,116 +50,197 @@ var parseManifest = function(content) {
+
+      // SUB Playlists
+      // The reflect is used so we can still continue running even if one of the playlists is broken.
+      async.map(playlists, async.reflect(function(p, cb) {
```
is this still needed with the newer error strategy?

mwhipple commented on src/walk-manifest.js:
```diff
@@ -49,116 +50,197 @@ var parseManifest = function(content) {
+          }
+        });
+
+        var flattenedResource = [].concat.apply([], results);
```
might be worth extracting this or creating a polyfill for `flat`

mwhipple replied:
if these arrays are being created above, why not just concat/flatten them there? Hopefully the reflected stuff can be washed away by more standard control flow stuff though.

mwhipple commented:
looks good to me, hopefully that control flow bit can just lean on existing machinery

mwhipple commented on test/unit/walk-manifest.spec.js:
```diff
@@ -153,8 +160,15 @@ describe('walk-manifest', function() {
       var errors = [];
-      var addErrorIntoErrors = function(err, resources, callback) {
-        errors.push(err);
+      var addErrorIntoErrors = function(err, uri, resources, callback) {
```
wrapping the object in a closure (similar to the previous example) would avoid the need to duplicate the function

mwhipple commented on src/walk-manifest.js:
```diff
@@ -202,9 +200,9 @@ var walkPlaylist = function(options, callback) {
+      async.map(playlists, function(p, cb) {
         if (!p.uri) {
-          return cb(null);
+          return cb(null, null);
```
might be nice to return `[]` instead of null (not sure if it matters with the code involved).

grsin commented on src/walk-manifest.js:
```diff
@@ -49,116 +50,180 @@ var parseManifest = function(content) {
+    if (response.statusCode !== 200) {
+      const keyError = new Error(response.statusCode + '|' + keyUri);
+      console.error(keyError);
+      return callback(keyError);
```
There is a todo for the key part. Currently the way it was originally written I think the key gets ignore if there is an error, so I don't pass that onError higher up;

BrandonOCasey commented on src/walk-manifest.js:
```diff
@@ -49,116 +50,180 @@ var parseManifest = function(content) {
+    return onError(manifestError, manifest.uri, resources, callback);
+  }
+
+  request({url: manifest.uri, timeout: 1500}, function(error, response, body) {
```
We should use `requestretry` for this since instead of adding another thing, since its used in `write-data`

BrandonOCasey commented on src/walk-manifest.js:
```diff
@@ -49,116 +50,180 @@ var parseManifest = function(content) {
+  }
+
+  // get the aes key
+  request({url: keyUri, encoding: null, timeout: 1500}, function(error, response, keyContent) {
```
We should use requestretry for this since instead of adding another thing, since its used in write-data

BrandonOCasey commented on src/walk-manifest.js:
```diff
@@ -49,116 +50,180 @@ var parseManifest = function(content) {
+
+  if (visitedUrls.includes(manifest.uri)) {
+    var manifestError = new Error('Trying to visit the same uri again; stuck in a cycle');
+    return onError(manifestError, manifest.uri, resources, callback);
```
Should this be an error? I would think that it could just be a warning, and we can just skip duplicate manifests.

grsin replied:
I think it would be a bad case for the downloader to try to download a bad/cycled manifest. I think we can leave it up to the caller add in they're own `onError()` which bypasses the cycle error if they want.

mwhipple replied:
An issue with the error behavior is that this code isn't testing just for cycles but any duplication (cycles could be tested with something similar to a stack which contains direct ancestry only but the current approach also provides some memoization style optimization). Cycles are an error but the same rendition playlist may be referenced multiple times in the master playlist which is not a cycle and is not necessarily an error. Skipping rather than erroring would likely be the simplest way to handle that concern.

grsin commented:
I originally finished/continued with the callback style since I was more familiar with it. I've updated to use `Promises` now. I don't know if there any major issue using native promises in walk and bluebird promises in write. Also I don't really know what version of node this project is supposed to be on, what style to follow, and converting everything to ES6. I'll leave all that for a different PR.

gkatsev commented on src/index.js:
```diff
@@ -3,10 +3,15 @@ var WriteData = require('./write-data');
-  console.log("Downloading additional data...");
-  return WriteData(options.decrypt, options.concurrency, resources);
+  var options = {decrypt: options.decrypt, basedir: options.output, uri: options.input};
+  WalkManifest(options)
```
this promise isn't returned

gkatsev replied:
@grsin hey, can you make a PR to fix this? Thanks

gkatsev commented on src/index.js:
```diff
@@ -3,10 +3,15 @@ var WriteData = require('./write-data');
+      console.log("Downloading additional data...");
+      return WriteData(options.decrypt, options.concurrency, resources);
+    })
+    .catch(function(err) {
```
this isn't necessary.

mwhipple replied:
The `keyError` handler could likely also be removed and just allow the error to bubble

mwhipple commented on src/walk-manifest.js:
```diff
@@ -49,116 +49,207 @@ var parseManifest = function(content) {
+    if (visitedUrls.includes(manifest.uri)) {
+      var manifestError = new Error('[WARN] Trying to visit the same uri again; skipping to avoid getting stuck in a cycle');
+      manifestError.uri = manifest.uri;
+      console.error(manifestError);
```
this now seems a roundabout way to log a message
